File: android/src/main/java/expo/modules/applemusic/HistoryRestClient.kt
package expo.modules.applemusic

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/** History-domain Apple Music REST (recently played, heavy rotation). */
internal class HistoryRestClient(
  private val transport: AppleMusicRestTransport,
) {
  suspend fun getRecentlyPlayed(musicUserToken: String): List<Map<String, Any?>> =
    withContext(Dispatchers.IO) {
      val json =
        transport.getJson(musicUserToken, "/v1/me/recent/played", mapOf("limit" to "10"))
      mapTopLevelResourceArray(json) { AppleMusicJsonMapper.mapRecentlyPlayed(it) }
    }

  suspend fun getRecentlyPlayedTracks(musicUserToken: String, limit: Int): List<Map<String, Any?>> =
    withContext(Dispatchers.IO) {
      val json =
        transport.getJson(
          musicUserToken,
          "/v1/me/recent/played/tracks",
          mapOf("limit" to limit.toString()),
        )
      mapTopLevelResourceArray(json) { AppleMusicJsonMapper.mapSong(it) }
    }

  suspend fun getHeavyRotation(musicUserToken: String, limit: Int): List<Map<String, Any?>> =
    withContext(Dispatchers.IO) {
      val json =
        transport.getJson(
          musicUserToken,
          "/v1/me/history/heavy-rotation",
          mapOf("limit" to limit.toString()),
        )
      mapTopLevelResourceArray(json) { AppleMusicJsonMapper.mapRecentResource(it) }
    }

  suspend fun getRecentlyPlayedStations(musicUserToken: String, limit: Int): List<Map<String, Any?>> =
    withContext(Dispatchers.IO) {
      val json =
        transport.getJson(
          musicUserToken,
          "/v1/me/recent/radio-stations",
          mapOf("limit" to limit.toString()),
        )
      mapTopLevelResourceArray(json) { AppleMusicJsonMapper.mapStation(it) }
    }

  suspend fun getRecentlyAdded(musicUserToken: String, limit: Int, offset: Int): List<Map<String, Any?>> =
    withContext(Dispatchers.IO) {
      val json =
        transport.getJson(
          musicUserToken,
          "/v1/me/library/recently-added",
          mapOf("limit" to limit.toString(), "offset" to offset.toString()),
        )
      mapTopLevelResourceArray(json) { AppleMusicJsonMapper.mapRecentResource(it) }
    }
}

File: src/rest/history-rest-client.ts
import {
  mapRecentlyPlayed,
  mapRecentResource,
  mapSong,
  mapStation,
} from '../mappers/apple-music-json-mapper';
import type { AppleMusicRestTransport } from './apple-music-rest-transport';
import { mapTopLevelResourceArray } from './rest-json';

/** History-domain Apple Music REST (recently played, heavy rotation). */
export class HistoryRestClient {
  constructor(private readonly transport: AppleMusicRestTransport) {}

  async getRecentlyPlayed(musicUserToken: string) {
    const json = await this.transport.getJson('/v1/me/recent/played', { limit: '10' }, musicUserToken);
    return mapTopLevelResourceArray(json.data, mapRecentlyPlayed);
  }

  async getRecentlyPlayedTracks(musicUserToken: string, limit: number) {
    const json = await this.transport.getJson(
      '/v1/me/recent/played/tracks',
      { limit: String(limit) },
      musicUserToken,
    );
    return mapTopLevelResourceArray(json.data, mapSong);
  }

  async getHeavyRotation(musicUserToken: string, limit: number) {
    const json = await this.transport.getJson(
      '/v1/me/history/heavy-rotation',
      { limit: String(limit) },
      musicUserToken,
    );
    return mapTopLevelResourceArray(json.data, mapRecentResource);
  }

  async getRecentlyPlayedStations(musicUserToken: string, limit: number) {
    const json = await this.transport.getJson(
      '/v1/me/recent/radio-stations',
      { limit: String(limit) },
      musicUserToken,
    );
    return mapTopLevelResourceArray(json.data, mapStation);
  }

  async getRecentlyAdded(musicUserToken: string, limit: number, offset: number) {
    const json = await this.transport.getJson(
      '/v1/me/library/recently-added',
      { limit: String(limit), offset: String(offset) },
      musicUserToken,
    );
    return mapTopLevelResourceArray(json.data, mapRecentResource);
  }
}
